### ISSUE TITLE
vmauth: Hashed password storage

### ISSUE THREAD

**Issue by jfolz** [VictoriaMetrics/VictoriaMetrics#5766]
### Is your feature request related to a problem? Please describe

This isn't really a feature request, but more of an inquiry whether a PR that implements this functionality would be accepted before I invest more time into it.

Context: We've recently started using vmauth with basic auth, and storing passwords as clear text makes me anxious. The amount of risk and my level of anxiety are probably completely out of balance. Mostly, I'm worried about the config file leaking.

### Describe the solution you'd like

My implementation would add a field `password_hash` (or similar) to the user config. I currently favor argon2 as hashing method, implemented in [golang.org/x/crypto/argon2](https://pkg.go.dev/golang.org/x/crypto/argon2) (which would need to be vendored like some other `x/crypto` libraries).
Argon2 is a state-of-the-art password hashing method with configurable runtime cost. While hashing with recommended parameters can take hundreds of milliseconds, it can be configured down to as little as a few microseconds. My limited knowledge about password hashing is that high runtime cost is important if you expect short and simple passwords. If you can ensure passwords are, e.g., hundreds of characters long and complex, then even a very fast hashing function should be fine. Users can choose how slow they want hashing to be, e.g. `echo -n password | argon2 saltsalt -id -t 2 -m 4` produces this string with parameters encoded:

```
$argon2id$v=19$m=16,t=2,p=1$c2FsdHNhbHQ$TsIHTcvap+XUROS0oUqyXZkXxMgQQ5dyc0PmE63P6/U
```

Whether to use clear text or hashed passwords would be up to the user. Low-risk endpoints can use clear text or low-cost hashing. High-risk/low-frequency endpoints could use more expensive hashing parameters.

### Describe alternatives you've considered

Learn to stop worrying and love clear text passwords.

**Comment by jfolz** [comment]
@hagen1778 interesting. That would provide some amount of separation from the actual config, which is probably changed more often than credentials. However, unless we do something fancy, we would still have to have an environment file with clear text passwords stored somewhere, correct?

**Comment by hagen1778** [comment]
> However, unless we do something fancy, we would still have to have an environment file with clear text passwords stored somewhere, correct?

Clear password needs to be stored somewhere. In k8s password is usually propagated to environment via [secrets](https://kubernetes.io/docs/concepts/configuration/secret/), so developer can't see its original value. It can be seen only when manually login into the pod or in secrets vault.

**Comment by nex1gen** [comment]
Hello everyone!
We have a case in which this feature will be very useful.
Installation of vmstorage and vmauth on virtual machines is done using our custom ansible role.
We generate vmauth config via jinja template, users and passwords load from HashiCorp Vault and store in variables during installation.
So, when coping rendered vmauth config to server, would like not to display passwords clearly. Because anyone who has access could see them

**Comment by jfolz** [comment]
@nex1gen Does this not prevent leaking secrets in logs?
```
- name: Some template task
  template:
    src: mytemplate
    dest: myfinishedfile
  no_log: true
  diff: no
```
`no_log: true` turns off logging and `diff: no` prevents the file showing up in diffs.

**Comment by nex1gen** [comment]
@jfolz Yes, u r right, but as a result, we will still have file with passwords on our servers

**Comment by jfolz** [comment]
@nex1gen I'm with you there 😄 
@hagen1778 I'll look into implementing this soon-ish then

**Comment by jfolz** [comment]
Looks like soon-ish is a brisk 15 or so months.

@f41gh7 sadly, I believe supporting hashed passwords won't be quite as simple as that. In particular, unique salts per usename would not be possible. After reading through the existing code today, I believe substantial changes would be necessary.

One thing strikes me as particularly odd: As far as I can tell, `vmauth` currently accepts multiple passwords per username, and each combination can have its own routing rules. If this is something that can be "fixed", then that would reduce the complexity somewhat, and potential performance pitfalls (requiring multiple hash computations per password per request) could be avoided.
